test/builtin_dataset_mocks.py

```diff
@@ -72,6 +71,24 @@ def prepare(self, home, config):
         available_file_names = {path.name for path in root.glob("*")}
         required_file_names = {resource.file_name for resource in self.dataset.resources(config)}
         missing_file_names = required_file_names - available_file_names
+        extra_file_names = available_file_names - required_file_names
+
+        # Some datasets need to provide already preprocessed data, for example if the preprocessing includes downloads.
+        # Such data will be included in the `extra_file_names` while the raw data will be in the `missing_file_names`.
+        # This detects these cases and removes the corresponding entries from the sets since the files are neither extra
+        # nor missing.
+        if missing_file_names:
+            for missing in missing_file_names.copy():
+                extra_candidate = missing.split(".", 1)[0]
+                if extra_candidate in extra_file_names:
+                    missing_file_names.remove(missing)
+                    extra_file_names.remove(extra_candidate)
```

Comment:
Is this entire thing equivalent to this?

```py
missing = set(missing.split(".")[0] for missing in missing_file_names)
not_actually_missing = missing & extra_file_names

missing -= not_actually_missing
extra_file_names -= not_actually_missing

# I guess we need to change `if missing_file_names` into `if_missing` below
```

Reply:
The logic behind it is correct, but `missing -= not_actually_missing` does not work. `not_actually_missing` contains the file names without suffixes, but `missing` has them with suffixes. Thus, we wouldn't remove anything here.

Reply:
> but missing has them with suffixes

No, `missing` is `missing_file_names` without the suffixes

Reply:
True, I missed that. In that case the error message is less expressive since it will now say "`foo` is missing" although it should say `"foo.tar" is missing`.

Reply:
If what we're doing is really equivalent to the snippet above, then I can't make sense out of it.

A = available
R = required
M = missing = R - A
E = extra = A - R

So the intersection between M and E is null, right?

Reply:
Correct. And that is why we perform this matching. For example

```py
A = {"foo", "bar"}
R = {"foo.zip", "spam.zip"}
M = R - A  # {"foo.zip", "spam.zip"}
E = A - R  # {"foo", "bar"}

missing = next(iter(M))  # foo.zip
extra_candidate = missing_candidate.split(".")[0]  # foo
extra_candidate in E  # True
M.remove(missing)  # {"spam.zip"}
E.remove(extra_candidate)  # {"bar"}
```